File: Computer4/3-jul_18/SOLUTIONS/lvl4_towering_towers.cpp
#include <bits/stdc++.h>
using namespace std;
const int MM = 1e6+2;
int n, h[MM]; stack<int> stk;
int main(){
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> n;
    for(int i=0; i<n; i++) cin >> h[i];
    for(int i=0; i<n; i++) {
        while(!stk.empty() && h[i] >= h[stk.top()]){
            stk.pop();
        }
        if(stk.empty()) cout << i << " ";
        else cout << i - stk.top() << " ";
        stk.push(i);
    }
}

File: Computer4/3-jul_18/SOLUTIONS/lvl4_towering_towers.java
import java.util.*;
import java.io.*;
public class homework {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{
		int n = readInt(), h[] = new int[n]; Stack<Integer> stk = new Stack();
		for(int i=0; i<n; i++) {
			h[i] = readInt();
		}
		for(int i=0; i<n; i++) {
			while(!stk.empty() && h[i] >= h[stk.peek()]) {
				stk.pop();
			}
			if(stk.empty()) System.out.print(i + " ");
			else System.out.print(i - stk.peek() + " ");
			stk.push(i);
		}
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}